src/Dotnet.Status.Web/GitHubUserTokenHandler.cs

```diff
@@ -0,0 +1,145 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.DotNet.Web.Authentication.GitHub;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace DotNet.Status.Web
+{
+    public class GitHubUserTokenHandler : AuthenticationHandler<UserTokenOptions>
+    {
+        private readonly GitHubClaimResolver _resolver;
+        private readonly IDataProtector _dataProtector;
+        private readonly ITokenRevocationProvider _revocation;
+
+        public GitHubUserTokenHandler(
+            GitHubClaimResolver resolver,
+            IOptionsMonitor<UserTokenOptions> options,
+            IDataProtectionProvider dataProtector,
+            ITokenRevocationProvider revocation,
+            ILoggerFactory logger,
+            UrlEncoder encoder,
+            ISystemClock clock) : base(options, logger, encoder, clock)
+        {
+            _resolver = resolver;
+            _dataProtector = dataProtector.CreateProtector("github-token");
+            _revocation = revocation;
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            string protectedToken = null;
+            if (Request.Headers.TryGetValue("Authorization", out var authHeader) &&
+                AuthenticationHeaderValue.TryParse(authHeader.ToString(), out var auth) &&
+                auth.Scheme == "Bearer")
+            {
+                protectedToken = auth.Parameter;
+            }
+
+            if (string.IsNullOrEmpty(protectedToken) &&
+                Request.Query.TryGetValue("token", out var tokenQuery))
+            {
+                protectedToken = tokenQuery;
+            }
+
+            if (string.IsNullOrEmpty(protectedToken))
+            {
+                return AuthenticateResult.Fail("Authentication required");
```

Comment:
>Fail [](start = 42, length = 4)

Is the an AuthenticateResult.Skip thing? This should probably not be "Fail" because no authentication was requested.

Reply:
Hrm... maybe! I'll try that. There is "NoResult"

---
In reply to: [344318984](https://github.com/dotnet/arcade-services/pull/756#discussion_r344318984) [](ancestors = 344318984)